os/fs/smartfs/smartfs_smart.c

```diff
@@ -1914,57 +1938,79 @@ int smartfs_rename(struct inode *mountpt, const char *oldrelpath, const char *ne
 		}
 
 		/* Okay, we will create newfilename in the newentry directory */
-
 		newparentdirsector = newentry.firstsector;
-		newfilename = oldfilename;
 	}
 
 	/* Test if the new parent directory is valid */
-
 	if (newparentdirsector != 0xFFFF) {
 		/* We can move to the given parent directory */
 
 		mode = oldentry.flags & SMARTFS_DIRENT_MODE;
 		type = oldentry.flags & SMARTFS_DIRENT_TYPE;
 
-		ret = smartfs_createentry(fs, newparentdirsector, newfilename, type, mode, &newentry, oldentry.firstsector, NULL);
-		if (ret != OK) {
-			goto errout_with_semaphore;
-		}
-
-		/* Now mark the old entry as inactive */
-
-		readwrite.logsector = oldentry.dsector;
-		readwrite.offset = 0;
-		readwrite.count = fs->fs_llformat.availbytes;
-		readwrite.buffer = (uint8_t *)fs->fs_rwbuffer;
-		ret = FS_IOCTL(fs, BIOC_READSECT, (unsigned long)&readwrite);
-		if (ret < 0) {
-			fdbg("Error %d reading sector %d data\n", ret, oldentry.dsector);
+		/* Find an available invalid/empty entry to write the new entry */
+		ret = smartfs_find_availableentry(fs, newparentdirsector, &f_newentry);
+		if (ret != OK && ret != -ENOENT) {
+			fdbg("find_availableentry encountered a problem, cannot find entry for writing\n");
 			goto errout_with_semaphore;
 		}
-		// bug fix for rename
-		tmp_pntr = (uint8_t *)&fs->fs_rwbuffer[oldentry.doffset];
-		tmp_flag = tmp_pntr[0];
-		tmp_flag |= (tmp_pntr[1] << 8);
 
-#if CONFIG_SMARTFS_ERASEDSTATE == 0xFF
-		tmp_flag &= ~SMARTFS_DIRENT_ACTIVE;
+		/* Check if it is a simple rename operation, i.e. in the same directory */
+		if (oldparentdirsector == newparentdirsector) {
```

Comment:
if they are same sector, we will overwrite old one, right? so below logic is correct for now, In my opinion.
```
ret = smartfs_finddirentry(fs, &oldentry, oldrelpath, &oldparentdirsector, &oldfilename);
...
ret = smartfs_finddirentry(fs, &newentry, newrelpath, &newparentdirsector, &newfilename);

#if CONFIG_SMARTFS_USE_SECTOR_BUFFER
if (oldparentdirsector == newparentdirsector) {
	overwrite_old_sector;
} else 
#endif
	{
		smartfs_find_availableentry();
		if (ret == -ENOENT) {
			smartfs_createentry();
		} else {
			if (ret != OK) {
				fdbg("error.....");
				goto error;
			}
			fvdbg("Entry found in parent sector : %d, offset : %d\n");
		}
		smartfs_writeentry();
		invalidate_entry(oldentry);
	}
}
```
How do you think?

Reply:
The logic is slightly different.
```
if (oldparentdirsector == newparentdirsector)
    if (CRC is not enabled && we can find an erased entry)
         Do not use old entry to overwrite
    else
         Overwrite old entry
else
     find_availableentry()
     if (not found)
          createentry()
     writeentry()
     invalidateentry()
```

Reply:
No, I don't agree. We can overwrite only below condition
```
if ((old == new) && (USE_SECTOR_BUFFER == TRUE))
```
otherwise we should find entry .
That's why I commented many time that find_availableentry should return first invalid entry if CONFIG_SMARTFS_USE_SECTOR_BUFFER is enabled, of course return emptry entry if CONFIG_SMARTFS_USE_SECTOR_BUFFER is not enabled.

Reply:
So we agree on using the same entry for overwriting when CRC is enabled.

When CRC is not enabled: -
Case 1 - We can find an erased entry in the sector (not invalidated)
My logic - use new entry
Your logic - use new entry
So we are at the same logic.

Case 2 - We can find an invalidated entry
My logic - use the same entry for overwriting
`Read the sector into rwbuffer
Make changes to the buffer
Write from the buffer back to the MTD`
So we used 1 new sector while writing to MTD because of changes in bits from 0 to 1
And only 1 MTD WRITE operation.
Old sector is released.

Your logic - Use the invalidated entry that is found
`Find_availableentry
read the parent sector into rwbuffer
Make changes to buffer contents
Write back buffer to MTD
read old entry into buffer
Mark flags invalid and write back to MTD`

So 1 new sector is used while writing to the invalidated entry because bits are changed from 0 to 1 (same as my logic)
But MTD_WRITE is called 2 times (one for writing entry, one for invalidating old entry)
So one extra MTD_WRITE that has to be journaled.

Case 3 - no available entry
Similar to case 2
My logic uses 1 new sector and 1 MTD_WRITE
Your logic will use 1 new sector but 2 MTD_WRITES

Reply:
```
ret = smartfs_finddirentry(fs, &oldentry, oldrelpath, &oldparentdirsector, &oldfilename);
...
ret = smartfs_finddirentry(fs, &newentry, newrelpath, &newparentdirsector, &newfilename);

#if CONFIG_SMARTFS_USE_SECTOR_BUFFER
if (oldparentdirsector == newparentdirsector) {
	overwrite_old_sector; // overwrite entry of source path.
} else 
#endif
	{
		smartfs_find_availableentry(); // This had better return entry. If crc enabled, first inactive entry, other wise first empty entry.
		if (ret == -ENOENT) {
			smartfs_createentry(); // this had better called inside of smartfs_find_availableentry. This will create  new sector for entry and return sector & offset.
		} else {
			if (ret != OK) {
				fdbg("error.....");
				goto error;
			}
			fvdbg("Entry found in parent sector : %d, offset : %d\n");
		}
		smartfs_writeentry();
		invalidate_entry(oldentry);
	}
}